bobtimus/src/loan.rs

```diff
@@ -44,42 +45,520 @@ pub struct LoanOffer {
+pub struct Collateralization {
     pub collateralization: Decimal,
-    /// Interest rate in percent
-    pub interest_rate: Decimal,
+    /// Interest to be added on top of the base interest rate for this term.
+    pub interest_mod: Decimal,
 }
 
+// TODO: Make sure that removing sat_per_vbyte is OK here
 #[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
 pub struct LoanRequest {
-    #[serde(with = "::elements::bitcoin::util::amount::serde::as_sat")]
-    pub collateral_amount: Amount,
-    collateral_inputs: Vec<Input>,
-    #[serde(with = "::elements::bitcoin::util::amount::serde::as_sat")]
-    fee_sats_per_vbyte: Amount,
-    borrower_pk: PublicKey,
+    pub principal_amount: LiquidUsdt,
+    pub collateral_amount: LiquidBtc,
+    pub collateral_inputs: Vec<Input>,
+    pub collateralization: Decimal,
+    pub borrower_pk: PublicKey,
     /// Loan term in days
     pub term: u32,
-    borrower_address: Address,
+    pub borrower_address: Address,
+}
+
+pub fn calculate_interest_rate(
+    borrower_term: u32,
+    borrower_collateralization: Decimal,
+    term_thresholds: &[Term],
+    collateralization_thresholds: &[Collateralization],
+    base_interest_rate: Decimal,
+) -> Result<Decimal> {
+    let mut term_interest_mod = Decimal::ZERO;
+    for term in term_thresholds {
+        if borrower_term >= term.days {
+            term_interest_mod = term.interest_mod;
+            continue;
+        }
+        break;
+    }
+
+    let mut collateralization_interest_mod = Decimal::ZERO;
+    for collateralization in collateralization_thresholds {
+        if borrower_collateralization >= collateralization.collateralization {
+            collateralization_interest_mod = collateralization.interest_mod;
+            continue;
+        }
+        break;
+    }
+
+    let interest_rate = base_interest_rate
+        .checked_add(term_interest_mod)
+        .context("Overflow due to addition")?
+        .checked_add(collateralization_interest_mod)
+        .context("Overflow due to addition")?;
+
+    Ok(interest_rate)
+}
+
+pub fn calculate_repayment_amount(
+    principal_amount: LiquidUsdt,
+    interest_percentage: Decimal,
+) -> Result<LiquidUsdt> {
+    let principal_amount = Decimal::from(principal_amount.as_satodollar());
+
+    let repayment_amount = principal_amount
+        .checked_add(
+            principal_amount
+                .checked_mul(interest_percentage)
+                .context("multiplication overflow")?,
+        )
+        .context("addition overflow")?;
+    let repayment_amount = LiquidUsdt::from_satodollar(
+        repayment_amount
+            .to_u64()
+            .context("decimal cannot be represented as u64")?,
+    );
+
+    Ok(repayment_amount)
+}
+
+pub fn calculate_liquidation_price(
+    repayment_amount: LiquidUsdt,
+    collateral_amount: LiquidBtc,
+) -> Result<LiquidUsdt> {
+    let repayment_amount = Decimal::from(repayment_amount.as_satodollar());
```

Comment:
Even though the calculation look relatively simple it would be worth it to add a comment mentioning the whole equation.